<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/12-treugolnoe-cislo-s-bolsim-kolicestvom-delitelei' style="text-decoration: none; color: #1a73e8;">
            12-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 12

Последовательность треугольных чисел образуется последовательным сложением натуральных чисел.

Например, седьмое треугольное число равно

$$
1 + 2 + 3 + 4 + 5 + 6 + 7 = 28
$$

Первые десять треугольных чисел:

```text
1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...
```

Делители первых семи треугольных чисел:

```text
1:  1
3:  1, 3
6:  1, 2, 3, 6
10: 1, 2, 5, 10
15: 1, 3, 5, 15
21: 1, 3, 7, 21
28: 1, 2, 4, 7, 14, 28
```

Число `28` является первым треугольным числом, у которого больше пяти делителей.

Необходимо найти **первое треугольное число**, у которого более `500` делителей.

Ответ: 76576500

## 1) Последовательный перебор треугольных чисел

В данном решении треугольные числа последовательно генерируются до тех пор, пока не встретится число, имеющее более `500` делителей.

Количество делителей вычисляется отдельной функцией `count_divs()`. Для этого достаточно перебрать делители только до квадратного корня числа. Если найден делитель `i`, то одновременно существует и парный делитель `num // i`, поэтому счётчик увеличивается сразу на два. Для полного квадрата квадратный корень учитывается отдельно.

### Алгоритм

1. Последовательно генерируются треугольные числа.
2. Для каждого числа вычисляется количество делителей.
3. Делители перебираются только до квадратного корня числа.
4. Для каждой найденной пары делителей счётчик увеличивается на два.
5. Если число является полным квадратом, квадратный корень учитывается отдельно.
6. При нахождении треугольного числа с количеством делителей больше `500` алгоритм завершается.

### Особенности

- Треугольные числа генерируются последовательно.
- Для поиска делителей используется перебор до квадратного корня.
- Каждая пара делителей учитывается одновременно.
- Проверка полного квадрата исключает двойной подсчёт квадратного корня.
- Временная сложность зависит от количества проверяемых треугольных чисел и составляет примерно `O(T · √N)`, где `T` — число сгенерированных треугольных чисел, а `N` — их величина.
- Дополнительная память составляет `O(1)`.

### Вывод

Перебор делителей только до квадратного корня значительно ускоряет вычисление их количества по сравнению с полным перебором. Тем не менее для каждого нового треугольного числа делители вычисляются заново, поэтому при больших ограничениях алгоритм выполняет достаточно много повторных операций.

In [8]:
from time import perf_counter # измерение времени выполнения алгоритма

In [9]:
def count_divs_1(num):
    if num == 1:
        return 1

    c = 0
    sqrt_num = num ** 0.5
    if sqrt_num % 1 == 0:
        c += 1

    for i in range(1, int(sqrt_num)):
        if num % i == 0:
            c += 2
    return c


def get_result_1() -> int:
    d = 1
    n = 0
    while True:
        n += d
        if count_divs_1(n) > 500:
            return n
        d += 1

start = perf_counter()
print(get_result_1())
end = perf_counter()

print(f'{end - start:.5f}')

76576500
1.79684


## 2) Использование формулы треугольного числа

Во втором решении используется математическое свойство треугольных чисел.

Треугольное число с номером `n` вычисляется по формуле:

$$
T_n=\frac{n(n+1)}{2}.
$$

Поскольку числа `n` и `n + 1` являются соседними, они всегда взаимно просты. Кроме того, одно из них обязательно чётное, поэтому деление на `2` можно выполнить заранее:

$$
T_n=
\begin{cases}
\frac{n}{2}(n+1), & \text{если } n \text{ чётное},\\[4pt]
n\frac{n+1}{2}, & \text{если } n \text{ нечётное}.
\end{cases}
$$

Полученные множители также остаются взаимно простыми, поэтому количество делителей треугольного числа равно произведению количества делителей каждого из них:

$$
d(T_n)=d(a)\cdot d(b).
$$

Количество делителей каждого множителя вычисляется той же функцией `count_divs_2()`, что и в предыдущем решении.

### Алгоритм

1. Последовательно перебираются номера треугольных чисел.
2. В зависимости от чётности `n` число `2` исключается из одного из множителей.
3. Для обеих частей вычисляется количество делителей.
4. Количество делителей треугольного числа находится как произведение полученных значений.
5. Если число делителей превышает `500`, вычисляется и возвращается соответствующее треугольное число.

### Отличия от предыдущего решения

- Треугольное число не строится до проверки количества делителей.
- Используется формула треугольного числа.
- Количество делителей вычисляется отдельно для двух взаимно простых множителей.
- Сохраняется прежняя функция подсчёта делителей.
- Размер чисел, для которых ищутся делители, становится значительно меньше.

### Вывод

Использование формулы треугольного числа позволяет отказаться от поиска делителей у самого треугольного числа. Благодаря разбиению на два взаимно простых множителя количество проверяемых делителей существенно уменьшается, что значительно ускоряет выполнение алгоритма без изменения способа их подсчёта.

In [10]:
def count_divs_2(num):
    if num == 1:
        return 1

    c = 0
    sqrt_num = num ** 0.5
    if sqrt_num % 1 == 0:
        c += 1

    for i in range(1, int(sqrt_num)):
        if num % i == 0:
            c += 2
    return c


def get_result_2() -> int:
    n = 1

    while True:
        if n % 2 == 0:
            left = n // 2
            right = n + 1
        else:
            left = n
            right = (n + 1) // 2

        divisors = count_divs_2(left) * count_divs_2(right)

        if divisors > 500:
            return n * (n + 1) // 2

        n += 1


start = perf_counter()
print(get_result_2())
end = perf_counter()

print(f'{end - start:.5f}')

76576500
0.04679


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **100 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [12]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(), number=100) / 100
    print(f'Алгоритм {i}: {tm:.5f}')

Алгоритм 1: 1.72734
Алгоритм 2: 0.04378


## Заключение

В ходе решения задачи были рассмотрены два различных подхода к поиску первого треугольного числа, имеющего более `500` делителей.

Первый алгоритм последовательно генерирует треугольные числа и для каждого из них вычисляет количество делителей перебором до квадратного корня. Несмотря на использование оптимального способа подсчёта делителей, каждое новое треугольное число проверяется полностью, что приводит к большому объёму вычислений.

Во втором решении используется формула треугольного числа

$$
T_n=\frac{n(n+1)}{2},
$$

а также свойство взаимной простоты соседних чисел. Вместо подсчёта делителей самого треугольного числа алгоритм вычисляет количество делителей двух меньших множителей, после чего перемножает полученные значения. Это позволяет значительно сократить объём вычислений.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `1.72734` с
- **Алгоритм 2:** `0.04378` с

Полученные результаты показывают, что использование математических свойств задачи позволяет добиться значительно большей производительности. Второй алгоритм оказался почти в **40 раз быстрее** первого, сохранив при этом ту же функцию подсчёта делителей. Это наглядно демонстрирует, что грамотно выбранная математическая модель зачастую даёт гораздо больший прирост производительности, чем локальная оптимизация отдельных частей программы.